#### [github notebook link](https://github.com/NBPub/Springboard/blob/main/8.3.3%20SQL%20Case%20Study/SQL_Country_Club.ipynb)

# Instructions

- `files/sqlite_db_pythonsqlite.db` - This is the database file you are looking to use python to write SQL queries against.
  - [db repository link](https://github.com/NBPub/Springboard/blob/main/8.3.3%20SQL%20Case%20Study/files/sqlite_db_pythonsqlite.db)
- `files/SQLTask Tier 2.sql` - This is a text file with the list of questions we are looking to answer via SQL queries on `sqlite_db_pythonsqlite.db`. 
  - SQL queries used for submission also added to this [file](https://github.com/NBPub/Springboard/blob/main/8.3.3%20SQL%20Case%20Study/files/SQLTasks%20Tier%202.sql)

# Tasks
## [Questions 1-9](#PHPMyAdmin-Interface)
**Brainstorming while server is unavailable. I'm going to answer them all here . . .**

## [Questions 10-13](#Python-SQLite3-Questions)
 - **Notebook code utilizes [python sqlite3](https://docs.python.org/3/library/sqlite3.html) to query databse**
 -  [SQL Queries](#SQL-Queries)
    - *list of queries used to answer question in SQL-styled markdown codeblocks (like 1-9 above)*

## 
## 
---

- Questions 1-9

### PHPMyAdmin Interface
   - [Springboard SQL server](https://sql.springboard.com/)
     - *server has been unavailable every time I've tried assignment, I should be fine with various SQL broswers. I'm currently using Adminer to help with a PostgreSQL instance*
     - I attempted to answer questions locally, and have listed code below, expand cell if collapsed

*used [DB Browser for SQLite](https://sqlitebrowser.org/) to explore database and test queries*

Q1: Some of the facilities charge a fee to members, but some do not.
Write a SQL query to produce a list of the names of the facilities that do.
```sql
SELECT name FROM Facilities WHERE membercost = 0;
```

Q2: How many facilities do not charge a fee to members?
```sql
SELECT COUNT(*) FROM Facilities WHERE membercost = 0;
```


Q3: Write an SQL query to show a list of facilities that charge a fee to members,
where the fee is less than 20% of the facility's monthly maintenance cost.
Return the facid, facility name, member cost, and monthly maintenance of the
facilities in question.
```sql
SELECT facid, name, membercost, monthlymaintenance
FROM Facilities
WHERE membercost < monthlymaintenance*0.2;
```


Q4: Write an SQL query to retrieve the details of facilities with ID 1 and 5.
Try writing the query without using the OR operator.
```sql
SELECT facid, name, membercost, monthlymaintenance
FROM Facilities
WHERE membercost < monthlymaintenance*0.2;
```

Q5: Produce a list of facilities, with each labelled as
'cheap' or 'expensive', depending on if their monthly maintenance cost is
more than $100. Return the name and monthly maintenance of the facilities
in question.

```sql
SELECT facid, name, membercost, monthlymaintenance
FROM Facilities
WHERE membercost < monthlymaintenance*0.2;
```
Q6: You'd like to get the first and last name of the last member(s)
who signed up. Try not to use the LIMIT clause for your solution. */
```sql
SELECT facid, name, membercost, monthlymaintenance
FROM Facilities
WHERE membercost < monthlymaintenance*0.2;
```

Q7: Produce a list of all members who have used a tennis court.
Include in your output the name of the court, and the name of the member
formatted as a single column. Ensure no duplicate data, and order by
the member name.
```sql
SELECT facid, name, membercost, monthlymaintenance
FROM Facilities
WHERE membercost < monthlymaintenance*0.2;
```

Q8: Produce a list of bookings on the day of 2012-09-14 which
will cost the member (or guest) more than $30. Remember that guests have
different costs to members (the listed costs are per half-hour 'slot'), and
the guest user's ID is always 0. Include in your output the name of the
facility, the name of the member formatted as a single column, and the cost.
Order by descending cost, and do not use any subqueries.
```sql
SELECT facid, name, membercost, monthlymaintenance
FROM Facilities
WHERE membercost < monthlymaintenance*0.2;
```

Q9: This time, produce the same result as in Q8, but using a subquery.
```sql
SELECT facid, name, membercost, monthlymaintenance
FROM Facilities
WHERE membercost < monthlymaintenance*0.2;
```

---

 - Questions 10-13 using SQLite file
 
## Python SQLite3 Questions

In [1]:
import sqlite3
con = sqlite3.connect("files/sqlite_db_pythonsqlite.db")

**Q10:** Produce a list of facilities with a total revenue less than 1000.
The output of facility name and total revenue, sorted by revenue. Remember
that there's a different cost for guests and members!

In [ ]:
# Facilities have two costs: "membercost" + "guestcost", index by "facid"
# Members index by "memid", 0 is guest.
# Bookings have facid, memid, and number of slots

# Facility Name | Total Revenue
# Revenue = Member Bookings * member cost + Guest Bookings * guest cost

In [ ]:
# Sum member bookings * member cost for each facility, add same for guests, then take total

In [15]:
query = '''
SELECT name, member_rev+guest_rev AS total_revenue FROM 
(SELECT facid, name,  SUM(slots)*membercost AS member_rev
FROM Bookings LEFT JOIN Facilities USING(facid)
WHERE memid > 0 GROUP BY facid)
INNER JOIN (
SELECT facid, SUM(slots)*guestcost AS guest_rev
FROM Bookings LEFT JOIN Facilities USING(facid)
WHERE memid = 0 GROUP BY facid) 
USING(facid) WHERE total_revenue < 1000
ORDER BY total_revenue
'''

In [16]:
cur = con.cursor()
cur.execute(query)
result = cur.fetchall()
result

[('Table Tennis', 180), ('Snooker Table', 240), ('Pool Table', 270)]

| name | total_revenue |
|---------|-------|
| Table Tennis | 180 |
| Snooker Table | 240 |
| Pool Table | 270 |

---

**Q11:** Produce a report of members and who recommended them in alphabetic surname,firstname order

In [ ]:
# member rows have memid reference in recommended column

In [2]:
query = '''
SELECT M.surname "Surname", M.firstname "Firstname",
 R.surname || ', ' || R.firstname AS "Recommending Member"
FROM Members M
LEFT JOIN Members R on R.memid=M.recommendedby 
WHERE M.memid > 0
ORDER BY Surname, Firstname
'''


| Surname | Firstname | Recommending Member |
|---------|-------|----------|
| Bader | Florence | Stibbons, Ponder |
| Bader | Anne | Stibbons, Ponder |
| ... | ... | ... |
| Worthington-Smyth | Henry | Smith, Tracy |

In [5]:
cur = con.cursor()
cur.execute(query)
result = cur.fetchall()

In [6]:
result

[('Bader', 'Florence', 'Stibbons, Ponder'),
 ('Baker', 'Anne', 'Stibbons, Ponder'),
 ('Baker', 'Timothy', 'Farrell, Jemima'),
 ('Boothe', 'Tim', 'Rownam, Tim'),
 ('Butters', 'Gerald', 'Smith, Darren'),
 ('Coplin', 'Joan', 'Baker, Timothy'),
 ('Crumpet', 'Erica', 'Smith, Tracy'),
 ('Dare', 'Nancy', 'Joplette, Janice'),
 ('Farrell', 'David', None),
 ('Farrell', 'Jemima', None),
 ('Genting', 'Matthew', 'Butters, Gerald'),
 ('Hunt', 'John', 'Purview, Millicent'),
 ('Jones', 'David', 'Joplette, Janice'),
 ('Jones', 'Douglas', 'Jones, David'),
 ('Joplette', 'Janice', 'Smith, Darren'),
 ('Mackenzie', 'Anna', 'Smith, Darren'),
 ('Owen', 'Charles', 'Smith, Darren'),
 ('Pinker', 'David', 'Farrell, Jemima'),
 ('Purview', 'Millicent', 'Smith, Tracy'),
 ('Rownam', 'Tim', None),
 ('Rumney', 'Henrietta', 'Genting, Matthew'),
 ('Sarwin', 'Ramnaresh', 'Bader, Florence'),
 ('Smith', 'Darren', None),
 ('Smith', 'Darren', None),
 ('Smith', 'Jack', 'Smith, Darren'),
 ('Smith', 'Tracy', None),
 ('Stibbons',

---

**Q12:** Find the facilities with their usage by member, but not guests

In [ ]:
# Bookings have facid, memid, and slots they booked

In [7]:
query = '''
SELECT name AS facility, SUM(slots) AS member_usage FROM Bookings
LEFT JOIN Facilities USING(facid)
WHERE memid > 0 GROUP BY facid ORDER BY member_usage DESC
'''

In [8]:
cur = con.cursor()
cur.execute(query)
result = cur.fetchall()
result

[('Badminton Court', 1086),
 ('Tennis Court 1', 957),
 ('Massage Room 1', 884),
 ('Tennis Court 2', 882),
 ('Snooker Table', 860),
 ('Pool Table', 856),
 ('Table Tennis', 794),
 ('Squash Court', 418),
 ('Massage Room 2', 54)]

| facility | member_usage |
|---------|-------|
| Badminton Court | 1086 |
| Tennis Court 1 | 957 |
| ... | ... |
| Massage Room 2 | 54 |

---

**Q13:** Find the facilities usage by month, but not guests

*provided answer for all facilities by month, and for each facility by month*

**Members Using All Facilities by Month**

In [13]:
query = '''
SELECT SUBSTRING(starttime,6,2) AS month, SUM(slots) AS member_usage 
FROM Bookings
WHERE memid > 0 GROUP BY month
'''

cur = con.cursor()
cur.execute(query)
result = cur.fetchall()
result

[('07', 1061), ('08', 2531), ('09', 3199)]

| month | member_usage |
|---------|-------|
| 07 | 1061 |
| 08 | 2531 |
| 09 | 3199 |

**Members Using Each Facility by Month**

In [10]:
query = '''
SELECT SUBSTRING(starttime,6,2) AS month, name AS facility, SUM(slots) AS member_usage 
FROM Bookings
LEFT JOIN Facilities USING(facid)
WHERE memid > 0 GROUP BY month, facid
'''

In [11]:
cur = con.cursor()
cur.execute(query)
result = cur.fetchall()
result

[('07', 'Tennis Court 1', 201),
 ('07', 'Tennis Court 2', 123),
 ('07', 'Badminton Court', 165),
 ('07', 'Table Tennis', 98),
 ('07', 'Massage Room 1', 166),
 ('07', 'Massage Room 2', 8),
 ('07', 'Squash Court', 50),
 ('07', 'Snooker Table', 140),
 ('07', 'Pool Table', 110),
 ('08', 'Tennis Court 1', 339),
 ('08', 'Tennis Court 2', 345),
 ('08', 'Badminton Court', 414),
 ('08', 'Table Tennis', 296),
 ('08', 'Massage Room 1', 316),
 ('08', 'Massage Room 2', 18),
 ('08', 'Squash Court', 184),
 ('08', 'Snooker Table', 316),
 ('08', 'Pool Table', 303),
 ('09', 'Tennis Court 1', 417),
 ('09', 'Tennis Court 2', 414),
 ('09', 'Badminton Court', 507),
 ('09', 'Table Tennis', 400),
 ('09', 'Massage Room 1', 402),
 ('09', 'Massage Room 2', 28),
 ('09', 'Squash Court', 184),
 ('09', 'Snooker Table', 404),
 ('09', 'Pool Table', 443)]

### SQL Queries


#### 10
```sql
SELECT name, member_rev+guest_rev AS total_revenue FROM 
(SELECT facid, name,  SUM(slots)*membercost AS member_rev
FROM Bookings LEFT JOIN Facilities USING(facid)
WHERE memid > 0 GROUP BY facid)
INNER JOIN (
SELECT facid, SUM(slots)*guestcost AS guest_rev
FROM Bookings LEFT JOIN Facilities USING(facid)
WHERE memid = 0 GROUP BY facid) 
USING(facid) WHERE total_revenue < 1000
ORDER BY total_revenue
```

#### 11
```sql
SELECT M.surname "Surname", M.firstname "Firstname",
 R.surname || ', ' || R.firstname AS "Recommending Member"
FROM Members M
LEFT JOIN Members R on R.memid=M.recommendedby 
WHERE M.memid > 0
ORDER BY Surname, Firstname
```

#### 12
```sql
SELECT name AS facility, SUM(slots) AS member_usage FROM Bookings
LEFT JOIN Facilities USING(facid)
WHERE memid > 0 GROUP BY facid ORDER BY member_usage DESC
```

#### 13
```sql
/* All facilities by month */
SELECT SUBSTRING(starttime,6,2) AS month, SUM(slots) AS member_usage 
FROM Bookings
WHERE memid > 0 GROUP BY month;
/* Each facility by month */
SELECT SUBSTRING(starttime,6,2) AS month, name AS facility, SUM(slots) AS member_usage 
FROM Bookings
LEFT JOIN Facilities USING(facid)
WHERE memid > 0 GROUP BY month, facid;
```